openhcl/virt_mshv_vtl/src/processor/tdx/tlb_flush.rs

```diff
@@ -172,16 +185,98 @@ impl UhProcessor<'_, TdxBacked> {
         }
     }
 }
+
+#[derive(Debug, Inspect)]
+pub(super) struct AtomicTlbRingBuffer {
+    /// The contents of the buffer.
+    #[inspect(hex, with = "|x| inspect::iter_by_index(x.iter())")]
+    buffer: Box<[AtomicU64; FLUSH_GVA_LIST_SIZE]>,
+    /// The number of GVAs that have been added over the lifetime of the VM.
+    gva_list_count: AtomicUsize,
+    /// The number of GVAs that have started being added to the list over the
+    /// lifetime of the VM.
+    in_progress_count: AtomicUsize,
+    /// A guard to ensure that only one thread is writing to the list at a time.
+    write_lock: Mutex<()>,
+}
+
+pub(super) struct AtomicTlbRingBufferWriteGuard<'a> {
+    buf: &'a AtomicTlbRingBuffer,
+    _write_lock: MutexGuard<'a, ()>,
+}
+
+impl AtomicTlbRingBuffer {
+    fn new() -> Self {
+        Self {
+            buffer: Box::new(std::array::from_fn(|_| AtomicU64::new(0))),
+            gva_list_count: AtomicUsize::new(0),
+            in_progress_count: AtomicUsize::new(0),
+            write_lock: Mutex::new(()),
+        }
+    }
+
+    fn count(&self) -> usize {
+        self.gva_list_count.load(Ordering::Acquire)
+    }
+
+    pub fn write(&self) -> AtomicTlbRingBufferWriteGuard<'_> {
+        let write_lock = self.write_lock.lock();
+        AtomicTlbRingBufferWriteGuard {
+            buf: self,
+            _write_lock: write_lock,
+        }
+    }
+
+    fn try_copy(&self, start_count: Wrapping<usize>, flush_addrs: &mut [u64]) -> bool {
+        let mut index = start_count;
+        for flush_addr in flush_addrs.iter_mut() {
+            *flush_addr = self.buffer[index.0 % FLUSH_GVA_LIST_SIZE].load(Ordering::Acquire);
+            index += 1;
+        }
+
+        // Check to see whether any additional entries have been added
+        // that would have caused a wraparound. If so, the local list is
+        // incomplete and the copy has failed.
+        if (Wrapping(self.in_progress_count.load(Ordering::Acquire)) - start_count).0
+            > FLUSH_GVA_LIST_SIZE
+        {
+            return false;
+        }
+        true
+    }
+}
+
+impl AtomicTlbRingBufferWriteGuard<'_> {
+    pub fn extend(&self, items: impl ExactSizeIterator<Item = HvGvaRange>) {
+        debug_assert_eq!(
+            self.buf.in_progress_count.load(Ordering::Acquire),
+            self.buf.gva_list_count.load(Ordering::Acquire)
+        );
+        // Adding a new item to the buffer must be done in three steps:
+        // 1. Indicate that an entry is about to be added so that any flush
+        //    code executing simultaneously will know that it might lose an
+        //    entry that it is expecting to see.
+        // 2. Add the entry.
+        // 3. Increment the valid entry count so that any flush code executing
+        //    simultaneously will know it is valid.
+        let len = items.len();
+        let count = self.buf.in_progress_count.fetch_add(len, Ordering::AcqRel);
```

Comment:
I think this can be a relaxed store:
```
let count = self.buf.in_progress_count.load(Relaxed).wrapping_add(len);
self.buf.in_progress_count.store(count, Relaxed);
```

Then keep the release in the loop, and switch to release on GVA list count.